core/src/main/java/org/apache/iceberg/view/ViewMetadata.java

```diff
@@ -337,95 +351,80 @@ public Builder removeProperties(Set<String> propertiesToRemove) {
-      this.changes.add(new MetadataUpdate.RemoveProperties(propertiesToRemove));
+      propertiesToRemove.forEach(properties::remove);
+      changes.add(new MetadataUpdate.RemoveProperties(propertiesToRemove));
       return this;
     }
 
-    private static Map<Integer, ViewVersion> indexVersions(List<ViewVersion> versionsToIndex) {
-      ImmutableMap.Builder<Integer, ViewVersion> builder = ImmutableMap.builder();
-      for (ViewVersion version : versionsToIndex) {
-        builder.put(version.versionId(), version);
-      }
-
-      return builder.build();
-    }
-
-    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
-      ImmutableMap.Builder<Integer, Schema> builder = ImmutableMap.builder();
-      for (Schema schema : schemasToIndex) {
-        builder.put(schema.schemaId(), schema);
-      }
-
-      return builder.build();
-    }
-
     public ViewMetadata build() {
       Preconditions.checkArgument(null != location, "Invalid location: null");
       Preconditions.checkArgument(versions.size() > 0, "Invalid view: no versions were added");
 
-      checkCurrentVersionIdIsValid(currentVersionId);
-
-      int versionHistorySizeToKeep =
+      int historySize =
           PropertyUtil.propertyAsInt(
               properties,
               ViewProperties.VERSION_HISTORY_SIZE,
               ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
 
       Preconditions.checkArgument(
-          versionHistorySizeToKeep > 0,
+          historySize > 0,
           "%s must be positive but was %s",
           ViewProperties.VERSION_HISTORY_SIZE,
-          versionHistorySizeToKeep);
-
-      if (versions.size() > versionHistorySizeToKeep) {
-        List<ViewVersion> versionsToKeep =
-            versions.subList(versions.size() - versionHistorySizeToKeep, versions.size());
-        List<ViewHistoryEntry> historyToKeep =
-            history.subList(history.size() - versionHistorySizeToKeep, history.size());
-        List<MetadataUpdate> changesToKeep = Lists.newArrayList(changes);
-        Set<MetadataUpdate.AddViewVersion> toRemove =
-            changesToKeep.stream()
-                .filter(update -> update instanceof MetadataUpdate.AddViewVersion)
-                .map(update -> (MetadataUpdate.AddViewVersion) update)
-                .filter(
-                    update ->
-                        update.viewVersion().versionId() != currentVersionId
-                            && !versionsToKeep.contains(update.viewVersion()))
-                .collect(Collectors.toSet());
-        changesToKeep.removeAll(toRemove);
-
-        versions = versionsToKeep;
-        history = historyToKeep;
-        changes = changesToKeep;
+          historySize);
+
+      // expire old versions, but keep at least the versions added in this builder
+      int numAddedVersions = (int) changes(MetadataUpdate.AddViewVersion.class).count();
+      int numVersionsToKeep = Math.max(numAddedVersions, historySize);
```

Comment:
This is a behavior change. If more versions were added than should be retained, at least one metadata file should contain them.